### PR TITLE
Improve speed, memory efficiency with alternate hashmap

### PR DESCRIPTION
**Pull request by jamesob**
**tl;dr**

I try swapping out `std::unordered_map` for a faster third-party implementation where it matters, see something like 15% speedup for initial block download coincident with a 19.3% reduction in memory usage.

---

When profiling initial block download, it's evident that a decent chunk of time on the critical path is spent in `CCoinsViewCache`, the data structure responsible for the in-memory representation of the UTXO set.

Selection_147
*Profile of the `msghand` thread at height ~300,000.*

The essential data member of the cache, `cacheCoins`, is a `std::unordered_map` that holds as many UTXO as can fit into memory given the `dbcache` setting. While this map implementation is definitely preferable to `std::map` (which retains order and has O(log(n)) lookups insted of O(1)), both stdlib maps are subpar [relative to non-`std` alternatives](https://martin.ankerl.com/2019/04/01/hashmap-benchmarks-01-overview/).

After seeing cacheCoins as a bottleneck, I decided to try swapping one of these alternative implementations in to see if there is any benefit. It looks like there is.

ibd local range 500000 520000

For 20,000 blocks at a representative part of the chain, we see something like a 15% speedup with a 19.3% reduction in memory usage (dbcache was set to 3000).

The `CCoinsCaching` benchmark shows improvement but not as dramatically as IBD does:

microbenches

Obviously adding 2000 lines of new consensus-critical code isn't appealing, but if those numbers hold for the full IBD process I think there might be a compelling case to be made for the added code.

### Implementations considered

Of the few best-looking options, I picked the one with the most succinct implementation and fewest dependencies, [`robin_hood::unordered_node_map`](https://github.com/martinus/robin-hood-hashing) (it's just a header file). I also looked at Facebook's [`F14NodeMap`](https://github.com/facebook/folly/blob/master/folly/container/F14.md) and Abseil's [`node_hash_map`](https://abseil.io/docs/cpp/guides/container) but both required too much elbow-grease (for now) to integrate here without pulling in a whole lot of code.

### Next steps

I'm going to be running more comprehensive benchmarks, including a near-full IBD from the network, to see if these improvements hold.

I'm also interested in benching this along with a similar change to `mapBlockIndex` (currently a `std::unordered_map`).

### Questions

If we decide to include this or another hashmap implementation, should we do a full fork inclusion a la leveldb or should we try to slim down the implementation to a single header file as is done here (but more)?

### CODE REVIEW

**Comment by promag** [comment]
> including a near-full IBD from the network, to see if these improvements hold.

Yap that would be interesting.

> I'm also interested in benching this along with a similar change to `mapBlockIndex`

I'd be surprised to see such improvements here.

> If we decide to include this or another hashmap implementation, should we do a full fork inclusion a la leveldb or should we try to slim down the implementation to a single header file as is done here (but more)?

Either way it should include tests.

**Comment by practicalswift** [comment]
@jamesob Interesting! Thanks for providing benchmarks!

Which implementation of `std::unordered_map` did you benchmark against?

It would be interesting to see this benchmarked against the three major implementations of `std::unordered_map` (`libstdc++`, `libc++` and Microsoft STL).

**Comment by jamesob** [comment]
> Which implementation of std::unordered_map did you benchmark against?

I've been benching on Ubuntu and Debian systems with `libstdc++` (x86_64, glibc6).

**Comment by fanquake** [comment]
> I try swapping out std::unordered_map for a faster third-party implementation where it matters, see something like 15% speedup for initial block download coincident with a 19.3% reduction in memory usage.

Nice. This is the kind of PR I enjoy reading / reviewing / testing.

I did a basic benchmark on my macOS machine: `time src/bitcoind -stopatheight=350000`.

#16718 6f9882ce4a817c9f14aa7526165ab6e278de890e - 48m49s
master db67101c748c208cced8e9b76a66d57cd48fbf6e - 55m22s

Will run something to a more intense height later.

**Comment by Sjors** [comment]
Concept ACK for improving  `std::unordered_map` behavior or swapping it for an alternative if the improvement is >10%. It's nice how this only changes 1 line in `coins.h`.

I'm able to reproduce the memory gain, but the speed is actually worse for me, on a 2019 MacBook Pro with `-dbcache=10000`, no indexes and no wallets, syncing over LAN to a single node:
* IBD before: 3:45 hours (memory at height 564180:  8.23 GB, plus 11 minute db flush) 
* IBD after: 5:30 hours (memory at height 564180: 7.5 GB, plus 20 minute db flush)

**Comment by martinus** [comment]
Hi all! I am the author of the the `robin_hood::unordered_map` variants, and just saw this PR by accident. Please ask away if you have any question about my implementation! 

I have also done extensive map benchmarks here, which might be interesting: https://github.com/martinus/map_benchmark

I agree with @JeremyRubin that the `throwOverflowError()` is a potential problem, if there is a way to force collisions in the map. Be aware that new hasmaps (facebook's F14, Google's Swisstable) also depend on good hashes. They don't throw an exception, but performance can degrade so much that it's not much different from an infinite loop.

File: src/test/coins_tests.cpp [hunk 1]
```diff
@@ -64,7 +64,7 @@ class CCoinsViewTest : public CCoinsView
                     map_.erase(it->first);
                 }
             }
-            mapCoins.erase(it++);
+            it = mapCoins.erase(it);
```
Comment (martinus) [inline src/test/coins_tests.cpp]: Be aware that there is a bug in robin_hood map https://github.com/martinus/robin-hood-hashing/issues/42 (as well as in tessil's map https://github.com/Tessil/robin-map/issues/20) where it is possible that entries are iterated over a second time before `end()` is reached.

**Comment by martinus** [comment]
@elichai, we have been using the map in several core components in [Dynatrace's](https://www.dynatrace.com) proprietary monitoring platform for quite a while. That's a very well tested and widely deployed product on multiple platforms (I work at dynatrace so I might be biased..).

I don't know much about other use cases. Here is a github query tough to find other uses of the map: https://github.com/search?q=robin_hood_h_included&type=Code It only finds usages who copied the file though.

Some usages that I had a look at: 
* https://github.com/jinja2cpp/Jinja2Cpp reported 2x speedup https://twitter.com/flex_ferrum/status/1159210516943908865
* https://github.com/carlfindahl/wacman 2x speedup: https://github.com/carlfindahl/wacman/commit/2189bef22eb86d4ad227dcd3b7ed1ca412110d15#diff-275e07678407ddce12c68ed11772f041R90
* https://github.com/masc-ucsc/lgraph serializes and mmaps robin_hood maps https://github.com/masc-ucsc/lgraph/commit/ec37bdc6e867824a40e269bd5204cd42b1a75cf3. I don't know of any speedups.
* https://github.com/p-ranav/csv has switched from tessil's robin_hood map to my map and reported speedup here: https://github.com/p-ranav/csv/commit/eabe6172cad9bad1f35b0749d729cbb3df89c8c2
* https://github.com/JeremyBYU/polylidarv2 has 30% speedup https://github.com/JeremyBYU/polylidarv2/commit/8e6e07bff1de5046cf60af6a62045abba7cf15ea

**Comment by laanwj** [comment]
The speed-ups are promising, it's surprising to me that this makes so much of a difference,

But to me it sounds quite risky, to be honest, to import a custom map implementation into consensus-critical code. Of course, we can never be sure, but likely these have much less testing than whatever tried-and-true map implementation comes with C++ libraries by default.

(Importing rust code for this reason seems a very bad idea to me, I'd be surprised if it helps performance at all with a FFI barrier there, and that makes it doubly risky)

**Comment by TheBlueMatt** [comment]
Right, pulling in Rust for something like this is way too early/aggressive. I'm curious if its possible to get similar speedups by taking a hatchet to robinhood so that we can cut it down to something thats more reviewable, since obviously thats the key criteria.

**Comment by martinus** [comment]
It would be very interesting where exactly the speedup is coming from. It might be possible to reproduce this speedup by staying with `std::unordered_map`. E.g. I see two possibilities for improvement:

* robin_hood map requires less hashing and less key comparisons than std::unordered_map. It might be possible to save a lot of hashing & comparisons by simply adding the precalculated hash value to the key (so instead of the `COutPoint` use e.g. `std::pair<COutPoint, size_t>`. This unfortunately will increase memory usage a bit (but not much, since the key/value pair is already quite large).

* I use a bulk allocator that never frees any memory, but reuses previously allocated memory. So once the cache has reached its peak size, no allocations will be necessary any more. This can also be achieved with a custom allocator for std::unordered_map.

* If the speedup comes from the robin_hood style memory layout, that's bad luck - can't be achieved with `std::unordered_map`.

**Comment by martinus** [comment]
> master ([a7be1cc](https://github.com/bitcoin/bitcoin/commit/a7be1cc92be4946c4f042bccd3a1b007657f3241)):
> ```
>         User time (seconds): 7735.97
>         Major (requiring I/O) page faults: 15004449
> ```
> 
> 2019-08-robinhood:
> ```
>         User time (seconds): 5598.92
>         Major (requiring I/O) page faults: 1684815
> ```

I find it a bit fishy that the major page faults with robinhood are 10x less. Could disk cache affect the benchmark? Maybe drop the cache before benchmarking, I think `sync; echo 3 > /proc/sys/vm/drop_caches` should work

**Comment by martinus** [comment]
I've benchmarked `bitcoind -reindex-chainstate -stopatheight=550000 -printtoconsole=0 -dbcache=5000` on my machine (Intel i7-8700, g++ 9.1, libstdc++), and can confirm a significant speedup with @jamesob's changes. For accurate reproducability I have disabled frequency scaling and turbo boost for my cpu (with [sudo pyperf system tune](https://pyperf.readthedocs.io/en/latest/)). Runtime with `std::unordered_map` is 6471 seconds, and 5339 seconds with robin_hood map, 21% faster.

I've modified `SaltedOutpointHasher` and added a custom equals function to count the number of hashes and equal operations performed. To my surprise, on my system `std::unordered_map` needs *less* hash calls. It looks like it caches the full hash. (10.7 vs. 13.2 billion hashes). I've benchmarked just the hashing and equals operations, and this should account for about 323 vs. 391 seconds of CPU total.

Next I'm going to try to use a bulk pool allocator for std::unordered_map and see how that goes.

**Comment by martinus** [comment]
I have tried to use `boost::pool_allocator` and `boost::fast_pool_allocator` for the `std::unordered_map`, but performance was *much* slower than `std::allocator`.

Then I adapted my [BulkPoolAllocator](https://github.com/martinus/robin-hood-hashing/blob/master/src/include/robin_hood.h#L338) to be usable for the `std::unordered_map`, and results are much better.

Benchmark results for `bitcoind -reindex-chainstate -stopatheight=550000 -printtoconsole=0 -dbcache=5000` on my i7-8700:

|                 | unordered_map | robin_hood | unordered_map + BulkPoolAllocator |
|-----------------|---------------|------------|---------------------------------|
| elapsed h:mm:ss |       1:47:51 |    1:28:59 |                         1:32:56 |
| speedup %       |            0% |      21.2% |                           16.1% |
| max RSS kbyte   |       7553716 |    6301540 |                         6950432 |
| max RSS %       |          100% |      83.4% |                           92.0% |

So most of the savings can also be achieved by staying with `std::unordered_map` and adding the BulkPoolAllocator. This should be much easier to integrate since it's less than 200 lines of code. Also, the allocator might make sense for other node-based containers as well.

**Comment by JeremyRubin** [comment]
I think we may also be able to eke out some additional performance by making the keys the sha256 salted hash of the COutpoint rather than the COutpoint itself, and instead of using siphash taking entropy from that salted hash (as done for cuckoocache).


At the expense of one extra sha256 on insert/find (and less siphash), we get to drop 4 of the key bytes and have cheaper hash re-evaluation. Furthermore (this is a little bit platform dependent, but maybe we can detect it or wrap the hash in an is_integral type) we can set __cache_hash_code in the map to false, and save another 8 bytes. Depending on padding issues we may save even more space.

**Comment by jamesob** [comment]
Thanks for running those benchmarks, @martinus. I also think we should try calling `cacheCoins.reserve(dbcache / avg_coinscacheentry_size)` somewhere during startup with both implementations to see how that affects things.

On the chance that we still want to consider using the `robin_hood` map, I've stripped about 500 lines out of `robinhood.h` (mostly relating to `unordered_flat_map` which I don't think we have much use for). I refer to this stripped down version as `robinhood-slim` in the benchmarks below (and it's now the tip of this branch).

---

I've been running benchmarks pretty continuously for the last few days and here's a summary of the SSD-based timings:

Selection_153

Unsurprisingly it seems that performance improvement is most pronounced with a large dbcache. Presumably this is because as the number of entries in cacheCoins grows, `std::unordered_map`'s performance starts to degrade, `robin_hood` starts to shine, or both.

I found it interesting that over numerous runs, a large dbcache (7000) ended up being slower across the board than a smaller dbcache (6000 or 3000).

In general, the weird timing disparities from these results give me pause. I'm benching on dedicated, non-virtualized hardware that is for no purpose other than benchmarking, and I'm dropping caches before runs (`sudo /sbin/swapoff -a; sudo /sbin/sysctl vm.drop_caches=3;`). Yet we see weird outliers like the 500-minute robinhood reindex at dbcache=5000 - which I didn't believe initially but then reproduced. Why would a dbcache=7000 run take longer than a dbcache=3000? Is the runtime of -reindex-chainstate somehow weirdly determined by the dbcache value during IBD?

I'm quite confused by some aspects of these results. 

Anyway, here's the memory story:

Selection_154

We reliably see significant memory savings from the robinhood map.

---

I also ran -reindex-chainstate to 550,000 on equivalent machines with HDDs. The disparity between master and this branch has been almost unbelievable, but I've reliably reproduced it even after doing machine crossover. The following table is characteristic of the -reindex-chainstates I've done on HDD:

| branch | dbcache | time | memory |
| ------ | ------- | ---- | ------ |
| robinhood | 5000 | 8:46:59 | 6.38 GB |
| master | 5000 | 21:32:58 (stopped early) | 7.47 GB |

Similarly, IBD to the same height with the same dbcache took 47 hours on master (!?) and 17 hours on this branch.

I'm just kind of assuming these results are bogus somehow, although I'm very confused by the fact that I can reliably reproduce them even after swapping machines, but maybe someone with a spinning disk can confirm or deny that there's a dramatic difference in performance between this branch and master.

**Comment by MarcoFalke** [comment]
Thanks for doing those additional benchmarks @jamesob. I think that confirms the finding by @Sjors that the sync is slowed down with a high dbcache in https://github.com/bitcoin/bitcoin/pull/16718#issuecomment-525818499

**Comment by jamesob** [comment]
> I think that confirms the finding by @Sjors that the sync is slowed down with a high dbcache in #16718 (comment)

Note though that I differ from Sjors' findings in that this branch was ~2x faster than master for dbcache=7000 (reindex-chainstate), whereas in his results this branch was slower than master for a large dbcache (10000) during IBD. 


I lightly suspect that @Sjors' slowness on this branch relative to master was due to some kind of IBD-inherent variability since on every other benchmark run I've seen, this branch performs better than master. 

Sjors' total IBD runtime (<5 hours) was also less than my reindex-chainstate runtime (up to 8 hours), which I attribute to his probably-much-better hardware.

**Comment by martinus** [comment]
In https://github.com/martinus/bitcoin/tree/2019-08-bulkpoolallocator I'm currently working on a PR where I've extracted (and completely rewrote...) the pool allocator so it is usable in std::unordered_map. Unfortunately the standard allocator API is trickier than I've expected, but I think I got it now. It's performance is still as advertised in https://github.com/bitcoin/bitcoin/pull/16718#issuecomment-526681466. I still have to clean up the code and add some unit tests before I'm creating a PR, probably in the next few days.

**Comment by luke-jr** [comment]
Surprised nobody's pointed out that using `std::unordered_map` basically gives us an *unknown* implementation right now... which probably is best avoided in consensus code, when possible. The fact that we can get performance improvements too really suggests to me that switching to a *specific* implementation would be a good idea. So concept ACK.

Agree with need for intense care and testing to avoid a surprise protocol change. Bundling like LevelDB seems best if it's a lot of code.

As far as issues with bad hashes - we don't want to enable intentionally-crafted transactions to screw over performance, so might need to re-hash with some seed?

**Comment by ryanofsky** [comment]
Status of this?

- Does using std::unordered_map with custom allocator https://github.com/bitcoin/bitcoin/pull/16718#issuecomment-527571982 still seem like it could be a good approach?

- Are the code changes in this PR more or less complete, or is there more robinhood code streamlining to be done?

- Are there still concerns about performance will smaller dbcaches and more constrained hardware with these implementations?

- Are there still benchmark reproducibility issues? Are these affecting other setups or just James' benchmarks?

**Comment by martinus** [comment]
> Status of this?

I think using a custom allocator from #16801 is a much safer choice than switching to a different map implementation.Writing a correct and fast hashmap is orders of magnitudes more difficult than a custom allocator. This PR was good though that it has highlighted what could be possible. Most of the performance benefit can be achieved with the custom allocator as well 

Concerning benchmark reproducibility, I believe @jamesob has stumbled upon a pretty severe performance bug, but we don't know yet if this is due to the new code or maybe the optimizations trigger an existing bug more often: https://github.com/bitcoin/bitcoin/pull/17060#issuecomment-544367528

**Comment by jamesob** [comment]
Some new (and stabilized) benchmarks here. Following the same methodology  [as in similar PRs](https://github.com/bitcoin/bitcoin/pull/16801#issuecomment-551883348), we see a 6% difference in runtime with ~400MB less memory usage for a reindex up to 550,000 ([`bench/robinhood.1`](https://github.com/jamesob/bitcoin/commits/bench/robinhood.1) vs. [`bench/master.1`](https://github.com/jamesob/bitcoin/commits/bench/master.1)):

```
host         tag                      time       time% maxmem  cpu%  dbcache
bench-ssd-2  robinhood.1              8:43:36    0.94  5304.37MB 356%  4000MB
bench-ssd-2  robinhood.1              8:44:29    0.94  5309.67MB 355%  4000MB
bench-ssd-2  master.1                 9:15:02    1.00  5776.95MB 344%  4000MB
bench-ssd-2  master.1                 9:15:26    1.00  5762.75MB 343%  4000MB

bench-ssd-3  robinhood.1              8:44:41    0.94  5293.84MB 355%  4000MB
bench-ssd-3  robinhood.1              8:46:29    0.94  5309.91MB 354%  4000MB
bench-ssd-3  master.1                 9:18:07    1.00  5780.84MB 342%  4000MB
bench-ssd-3  master.1                 9:15:53    1.00  5773.25MB 343%  4000MB

bench-ssd-4  robinhood.1              8:46:13    0.95  5279.41MB 354%  4000MB
bench-ssd-4  robinhood.1              8:45:36    0.94  5294.21MB 354%  4000MB
bench-ssd-4  master.1                 9:14:42    1.00  5782.72MB 344%  4000MB
bench-ssd-4  master.1                 9:16:17    1.00  5762.92MB 343%  4000MB

bench-ssd-5  robinhood.1              8:42:02    0.94  5277.88MB 357%  4000MB
bench-ssd-5  robinhood.1              8:41:34    0.94  5302.08MB 357%  4000MB
bench-ssd-5  master.1                 9:12:43    1.00  5773.20MB 345%  4000MB
bench-ssd-5  master.1                 9:12:08    1.00  5769.65MB 345%  4000MB

bench-strong robinhood.1              14:16:20   0.94  5302.65MB 563%  4000MB
bench-strong master.1                 15:07:08   1.00  5804.43MB 543%  4000MB
bench-strong master.1                 15:08:44   1.00  5722.19MB 543%  4000MB
```

Interesting, but probably not a sufficient benefit to justify adding an extra 1500 lines of consensus-critical code. I'm going to close this one and perhaps someone can revisit this later.
